File: python/dsa/heaps/binary_heap.py
# Heaps — Binary Min-Heap

"""Array-based binary min-heap.

insert/extract_min/decrease_key O(log n); heapify O(n); peek O(1).
"""

from __future__ import annotations

from typing import Generic, Iterable, TypeVar

T = TypeVar("T")


class BinaryHeap(Generic[T]):
    """Binary min-heap backed by a dynamic array."""

    def __init__(self) -> None:
        """Creates an empty heap.

        Time: O(1); Space: O(1).
        """
        self._heap: list[T] = []

    def size(self) -> int:
        """Returns the number of elements.

        Time: O(1); Space: O(1).
        """
        return len(self._heap)

    def peek(self) -> T:
        """Returns the minimum without removing.

        Time: O(1); Space: O(1).
        """
        if not self._heap:
            raise RuntimeError("Heap is empty")
        return self._heap[0]

    def insert(self, value: T) -> None:
        """Inserts a value.

        Time: O(log n); Space: O(1).
        """
        self._heap.append(value)
        self._sift_up(len(self._heap) - 1)

    def extract_min(self) -> T:
        """Removes and returns the minimum.

        Time: O(log n); Space: O(1).
        """
        if not self._heap:
            raise RuntimeError("Heap is empty")
        minimum = self._heap[0]
        last = self._heap.pop()
        if self._heap:
            self._heap[0] = last
            self._sift_down(0)
        return minimum

    def decrease_key(self, index: int, new_value: T) -> None:
        """Decreases the key at index to new_value (must be smaller).

        Time: O(log n); Space: O(1).
        """
        if index < 0 or index >= len(self._heap):
            raise IndexError(index)
        if new_value > self._heap[index]:
            raise ValueError("New value must not be greater")
        self._heap[index] = new_value
        self._sift_up(index)

    def index_of(self, value: T) -> int:
        """Returns the index of an element, or -1.

        Time: O(n); Space: O(1).
        """
        try:
            return self._heap.index(value)
        except ValueError:
            return -1

    def heapify(self, items: Iterable[T]) -> None:
        """Builds a min-heap from items in place on the internal array.

        Time: O(n); Space: O(n).
        """
        self._heap = list(items)
        for i in range(self._parent(len(self._heap) - 1), -1, -1):
            self._sift_down(i)

    def _sift_up(self, index: int) -> None:
        while index > 0:
            parent = self._parent(index)
            if self._heap[index] >= self._heap[parent]:
                break
            self._swap(index, parent)
            index = parent

    def _sift_down(self, index: int) -> None:
        n = len(self._heap)
        while True:
            left = self._left_child(index)
            right = self._right_child(index)
            smallest = index
            if left < n and self._heap[left] < self._heap[smallest]:
                smallest = left
            if right < n and self._heap[right] < self._heap[smallest]:
                smallest = right
            if smallest == index:
                break
            self._swap(index, smallest)
            index = smallest

    @staticmethod
    def _parent(i: int) -> int:
        return (i - 1) // 2

    @staticmethod
    def _left_child(i: int) -> int:
        return 2 * i + 1

    @staticmethod
    def _right_child(i: int) -> int:
        return 2 * i + 2

    def _swap(self, i: int, j: int) -> None:
        self._heap[i], self._heap[j] = self._heap[j], self._heap[i]

File: java/src/main/java/com/dsa/examples/heaps/BinaryHeap.java
// Heaps — Binary Min-Heap

package com.dsa.examples.heaps;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Array-based binary min-heap.
 * insert/extractMin/decreaseKey O(log n); heapify O(n); peek O(1).
 */
public class BinaryHeap<T extends Comparable<T>> {

    private final List<T> heap = new ArrayList<>();

    /**
     * Creates an empty heap.
     * Time: O(1); Space: O(1).
     */
    public BinaryHeap() {
    }

    /**
     * Returns the number of elements.
     * Time: O(1); Space: O(1).
     */
    public int size() {
        return heap.size();
    }

    /**
     * Returns the minimum without removing.
     * Time: O(1); Space: O(1).
     */
    public T peek() {
        if (heap.isEmpty()) {
            throw new IllegalStateException("Heap is empty");
        }
        return heap.get(0);
    }

    /**
     * Inserts a value.
     * Time: O(log n); Space: O(1).
     */
    public void insert(T value) {
        heap.add(value);
        siftUp(heap.size() - 1);
    }

    /**
     * Removes and returns the minimum.
     * Time: O(log n); Space: O(1).
     */
    public T extractMin() {
        if (heap.isEmpty()) {
            throw new IllegalStateException("Heap is empty");
        }
        T min = heap.get(0);
        T last = heap.remove(heap.size() - 1);
        if (!heap.isEmpty()) {
            heap.set(0, last);
            siftDown(0);
        }
        return min;
    }

    /**
     * Decreases the key at {@code index} to {@code newValue} (must be smaller).
     * Time: O(log n); Space: O(1).
     */
    public void decreaseKey(int index, T newValue) {
        if (index < 0 || index >= heap.size()) {
            throw new IndexOutOfBoundsException(index);
        }
        if (newValue.compareTo(heap.get(index)) > 0) {
            throw new IllegalArgumentException("New value must not be greater");
        }
        heap.set(index, newValue);
        siftUp(index);
    }

    /**
     * Returns the index of an element (for decreaseKey), or -1.
     * Time: O(n); Space: O(1).
     */
    public int indexOf(T value) {
        return heap.indexOf(value);
    }

    /**
     * Builds a min-heap from {@code items} in place on the internal array.
     * Time: O(n); Space: O(n).
     */
    public void heapify(List<T> items) {
        heap.clear();
        heap.addAll(items);
        for (int i = parent(heap.size() - 1); i >= 0; i--) {
            siftDown(i);
        }
    }

    /**
     * Builds a min-heap from an array.
     * Time: O(n); Space: O(n).
     */
    @SafeVarargs
    public final void heapify(T... items) {
        heapify(Arrays.asList(items));
    }

    private void siftUp(int index) {
        while (index > 0) {
            int p = parent(index);
            if (heap.get(index).compareTo(heap.get(p)) >= 0) {
                break;
            }
            swap(index, p);
            index = p;
        }
    }

    private void siftDown(int index) {
        int n = heap.size();
        while (true) {
            int left = leftChild(index);
            int right = rightChild(index);
            int smallest = index;
            if (left < n && heap.get(left).compareTo(heap.get(smallest)) < 0) {
                smallest = left;
            }
            if (right < n && heap.get(right).compareTo(heap.get(smallest)) < 0) {
                smallest = right;
            }
            if (smallest == index) {
                break;
            }
            swap(index, smallest);
            index = smallest;
        }
    }

    private static int parent(int i) {
        return (i - 1) / 2;
    }

    private static int leftChild(int i) {
        return 2 * i + 1;
    }

    private static int rightChild(int i) {
        return 2 * i + 2;
    }

    private void swap(int i, int j) {
        T t = heap.get(i);
        heap.set(i, heap.get(j));
        heap.set(j, t);
    }
}
